src/bigint.rs

```diff
@@ -467,146 +546,352 @@ impl Unsigned for BigUint {}
 
 forward_all_binop_to_val_ref_commutative!(impl Add for BigUint, add);
 
+fn add2(a: &mut [BigDigit], b: BigUintS) {
+    let mut b_iter = b.iter();
+    let mut carry = 0;
+
+    for ai in a.iter_mut() {
+        if let Some(bi) = b_iter.next() {
+            *ai = adc(*ai, *bi, &mut carry);
+        } else if carry != 0 {
+            *ai = adc(*ai, 0, &mut carry);
+        } else {
+            break;
+        }
+    }
+
+    debug_assert!(carry == 0);
+}
+
 impl<'a> Add<&'a BigUint> for BigUint {
     type Output = BigUint;
 
-    fn add(self, other: &BigUint) -> BigUint {
-        let mut sum = self.data;
+    fn add(mut self, other: &BigUint) -> BigUint {
+        let new_len = cmp::max(self.data.len(), other.data.len()) + 1;
```

Comment:
Hmm, this *always* grows `self.data`, even when the extra digit might not be needed.  It may be that the vector often has that extra capacity anyway, but... can we keep the property that the extra carry is only pushed when we know it really overflowed?  Maybe let `add2` return the leftover carry, for instance.